mu4/domain/importexport/internal/braille/exportbraille.cpp

```diff
@@ -0,0 +1,2324 @@
+            currentCleffType.push_back(ClefType::INVALID);
+            currentKey.push_back(Key::INVALID);
+            }
+        }
+    
+    void write(QIODevice* dev);
+    };
+    
+    //---------------------------------------------------------
+    //   saveBraille
+    //    return false on error
+    //---------------------------------------------------------
+    
+    /**
+     Save Score as Braille (brf) file \a name.
+     
+     Return false on error.
+     */
+    
+bool saveBraille(Score* score, QIODevice* device)
+    {
+    ExportBraille eb(score);
+    eb.write(device);
+    return true;
+    }
+    
+bool saveBraille(Score* score, const QString& name)
+    {
+    QFile f(name);
+    if (!f.open(QIODevice::WriteOnly))
+        return false;
+
+    bool res = saveBraille(score, &f) && (f.error() == QFile::NoError);
+    f.close();
+    return res;
+    }
+    
+void ExportBraille::resetOctaves()
+    {
+    for (int i = 0; i < score->staves().size(); ++i)
+        currentOctave[i] = -1;
+    }
+
+void ExportBraille::credits(QIODevice* dev)
+    {
+    QTextStream out(dev);
+    // find the vboxes in every page and write their elements as credit-words
+    for (const auto page : score->pages()) {
+        for (const auto system : page->systems()) {
+            for (const auto mb : system->measures()) {
+                if (mb->isVBox()) {
+                    for (const Element* element : mb->el()) {
+                        if (element->isText()) {
+                            const Text* text = toText(element);
+                            out << TextToUEBBraille().braille(text->plainText()).toUtf8() << endl;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        
+    QStringList creators;
+    // the creator types commonly found in MusicXML
+    creators << "arranger" << "composer" << "lyricist" << "poet" << "translator";
+    for (QString type : creators) {
+        QString creator = score->metaTag(type);
+        if (!creator.isEmpty())
+            out << TextToUEBBraille().braille(QString("%1 %2").arg(type).arg(creator)).toUtf8() << endl;
+        }
+    if (!score->metaTag("copyright").isEmpty())
+        out << TextToUEBBraille().braille(QString("© %2").arg(score->metaTag("copyright"))).toUtf8() << endl;
+    out << endl;
+    out.flush();
+    }
+    
+void ExportBraille::instruments(QIODevice* dev)
+    {
+    //Print staff number to instrument mapping.
+    QTextStream out(dev);
+    for (int i = 0; i < score->staves().size(); ++i) {
+        out << TextToUEBBraille().braille(QString("%1 %2").arg(i+1).arg(score->staves()[i]->part()->instrumentName())) << endl;
+        }
+    out << endl;
+    out.flush();
+    }
+    
+void ExportBraille::write(QIODevice* dev)
+    {
+    credits(dev);
+    instruments(dev);
+    int nrStaves = score->staves().size();
+    QString measureBraille[nrStaves];
+    QString line[nrStaves];
```

Comment:
These lines cause build errors with MSVC. See https://stackoverflow.com/a/33423538

Reply:
Probably best to use a `std::vector` of QStrings.

Reply:
@shoogle thank you for the pointer! I'll convert them into dynamic calls. MinGW was happy with them.

Reply:
It indeed does not

Reply:
Search the source for `VLA` to find how this is being worked around. Example:

    #if (!defined (_MSCVER) && !defined (_MSC_VER))
          char txt[size+1];
    #else
          // MSVC does not support VLA. Replace with std::vector. If profiling determines that the
          //    heap allocation is slow, an optimization might be used.
          std::vector<char> vtxt(size+1);
          char* txt = vtxt.data();
    #endif

Reply:
Addressed in b0365cb. Thank you! 👍 I used directly the `std::vector<QString> line(nrStaves);` to not have diverging code depending on the compiler if one code can accommodate all. Let's see what the build says.